server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -594,4 +609,48 @@ public void clear() {
             runnables.clear();
         }
     }
+
+    /**
+     * Verifies that the given query does not exceed the maximum allowed clause count.
+     * Traverses the query upfront to estimate its total cost and fails fast by throwing
+     * {@link TooManyNestedClauses} if the limit is exceeded.
+     */
+    private static void verifyQueryLimit(Query query) {
+        final int[] numClauses = new int[1];
+        query.visit(new QueryVisitor() {
+            @Override
+            public QueryVisitor getSubVisitor(BooleanClause.Occur occur, Query parent) {
+                // Return this instance even for MUST_NOT and not an empty QueryVisitor
+                return this;
+            }
+
+            @Override
+            public void visitLeaf(Query query) {
+                if (numClauses[0] > getMaxClauseCount()) {
+                    throw new TooManyNestedClauses();
+                }
+                ++numClauses[0];
+            }
+
+            @Override
+            public void consumeTerms(Query query, Term... terms) {
+                if (numClauses[0] > getMaxClauseCount()) {
```

Comment:
Maybe pull this into a variable at the top of this method so we don't call `getMaxClauseCount()` for every query?  I guess the JVM will optimize it away in most cases but nicer to be explicit about it.